Mathlib/Topology/ContinuousFunction/CompactlySupported.lean

```diff
@@ -0,0 +1,714 @@
+    NonUnitalNonAssocRing C_c(α, β) :=
+  DFunLike.coe_injective.nonUnitalNonAssocRing _ coe_zero coe_add coe_mul coe_neg coe_sub
+    (fun _ _ => rfl) fun _ _ => rfl
+
+instance instNonUnitalRing [NonUnitalRing β] [TopologicalRing β] : NonUnitalRing C_c(α, β) :=
+  DFunLike.coe_injective.nonUnitalRing _ coe_zero coe_add coe_mul coe_neg coe_sub (fun _ _ => rfl)
+    fun _ _ => rfl
+
+instance instNonUnitalCommRing [NonUnitalCommRing β] [TopologicalRing β] :
+    NonUnitalCommRing C_c(α, β) :=
+  DFunLike.coe_injective.nonUnitalCommRing _ coe_zero coe_add coe_mul coe_neg coe_sub
+    (fun _ _ => rfl) fun _ _ => rfl
+
+instance instIsScalarTower {R : Type*} [Semiring R] [NonUnitalNonAssocSemiring β]
+    [TopologicalSemiring β] [Module R β] [ContinuousConstSMul R β] [IsScalarTower R β β] :
+    IsScalarTower R C_c(α, β) C_c(α, β) where
+  smul_assoc r f g := by
+    ext
+    simp only [smul_eq_mul, coe_mul, coe_smul, Pi.mul_apply, Pi.smul_apply]
+    rw [← smul_eq_mul, ← smul_eq_mul, smul_assoc]
+
+instance instSMulCommClass {R : Type*} [Semiring R] [NonUnitalNonAssocSemiring β]
+    [TopologicalSemiring β] [Module R β] [ContinuousConstSMul R β] [SMulCommClass R β β] :
+    SMulCommClass R C_c(α, β) C_c(α, β) where
+  smul_comm r f g := by
+    ext
+    simp only [smul_eq_mul, coe_smul, coe_mul, Pi.smul_apply, Pi.mul_apply]
+    rw [← smul_eq_mul, ← smul_eq_mul, smul_comm]
+
+end AlgebraicStructure
+
+section ZeroAtInfty
+
+open ZeroAtInfty
+
+variable [TopologicalSpace β] [TopologicalSpace γ] [Zero γ]
+variable [FunLike F β γ] [CompactlySupportedContinuousMapClass F β γ]
+
+lemma zero_at_infty_of_hasCompactSupport (f : F) :
+    Filter.Tendsto f (Filter.cocompact β) (𝓝 0) := by
+  rw [_root_.tendsto_nhds]
+  intro s _ hzero
+  rw [Filter.mem_cocompact]
+  use tsupport f
+  constructor
+  · exact hasCompactSupport f
+  · intro x hx
+    simp only [Set.mem_preimage]
+    rw [← Set.not_mem_compl_iff, compl_compl] at hx
+    rw [image_eq_zero_of_nmem_tsupport hx]
+    exact hzero
+
+instance : ZeroAtInftyContinuousMapClass C_c(β, γ) β γ where
+  map_continuous f := f.continuous_toFun
+  zero_at_infty f := zero_at_infty_of_hasCompactSupport f
+
+end ZeroAtInfty
+
+section Uniform
+
+variable [UniformSpace β] [UniformSpace γ] [Zero γ]
+variable [FunLike F β γ] [CompactlySupportedContinuousMapClass F β γ]
+
+theorem uniformContinuous (f : F) : UniformContinuous (f : β → γ) :=
+  (map_continuous f).uniformContinuous_of_tendsto_cocompact (zero_at_infty_of_hasCompactSupport f)
+
+end Uniform
+
+/-! ### Metric structure
+
+When `β` is a metric space, then every element of `C_c(α, β)` is bounded, and so there is a natural
+inclusion map `CompactlySupportedContinuousMap.toBCF : C_c(α, β) → (α →ᵇ β)`. Via this map
+`C_c(α, β)` inherits a metric as the pullback of the metric on `α →ᵇ β`. Moreover, this map has
+closed range in `α →ᵇ β` and consequently `C_c(α, β)` is a complete space whenever `β` is complete.
```

Comment:
I don't think that's true: the closure of functions with compact support is rather functions tending to zero at infinity, so `C_c(α, β)` is not complete. And in fact the topology you're putting here is not the "right" topology on `C_c(α, β)`, so this metric structure should probably not be registered as an instance. Maybe drop this section from the PR, and come back to it later as there is some design to be carefully thought through?

Reply:
OK. I made another PR #12629 where I defined a subtype of compactly supported functions. I agree that `C_c` should be given a different topology, maybe the one similar to that for test functions? I removed all the part referring to metric.